src/sage/calculus/calculus.py

```
@@ -1410,75 +1492,178 @@ def limit(ex, dir=None, taylor=False, algorithm='maxima', **argv):
+    sage: limit(x**2 + y, "y", x) # Need y=var('y') defined for this test
+    x^2 + x
+
+    # Test conversion of a if not symbolic
+    sage: limit(x**2, x, "3")
+    9
+
+    # Test using a constant number as variable 'v' fails
+    sage: limit(x**2 + 5, 5, 10)
+    Traceback (most recent call last):
+    ...
+    TypeError: limit variable must be a variable, not a constant
     """
+    # Process expression
     if not isinstance(ex, Expression):
-        ex = SR(ex)
+        try:
+            ex = SR(ex)
+        except TypeError:
+            raise TypeError(f"Cannot convert expression to symbolic ring: {ex}")
+
+    # Argument parsing: Determining v and a based on syntax used
+    v = None
+    a = None
+
+    if len(args) == 2: # Syntax: limit(ex, v, a, ...)
+        if kwargs: # Cannot mix positional v, a with keyword args
+            raise ValueError(f"Use either limit(expr, v, a, ...) or limit(expr, v=a, ...) syntax. "
+                             f"Cannot mix positional v, a with keyword variable arguments: {list(kwargs.keys())}")
+        v = args[0]
+        a = args[1]
+    elif len(args) == 1:
+        if kwargs: 
+             raise ValueError(f"Use either limit(expr, v, a, ...) or limit(expr, v=a, ...) syntax. "
+                              f"Cannot mix positional v with keyword variable arguments: {list(kwargs.keys())}")
+        else:
+             raise ValueError("Use either limit(expr, v, a, ...) or limit(expr, v=a, ...) syntax. "
+                              f"Need three positional arguments (expr, v, a) or one keyword argument (v=a). "
+                              f"Got one positional argument for v, a.")
+    elif len(args) == 0: # Potential syntax: limit(ex, v=a, ...) or limit(ex)
+        if len(kwargs) == 1: 
+            k, = kwargs.keys()
+            if not isinstance(k, str):
+                 raise ValueError(f"Invalid variable specification in keyword argument: {k} (must be a string)")
+            try:
```

no need to guard with try. Just let error propagate.